File: app/src/main/java/io/github/hyuwah/catatanku/NoteListActivity.java
package io.github.hyuwah.catatanku;

import android.app.AlertDialog;
import android.app.LoaderManager;
import android.app.SearchManager;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.Context;
import android.content.CursorLoader;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.Loader;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.support.customtabs.CustomTabsIntent;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.ActionMode;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.AbsListView;
import android.widget.ListView;
import android.widget.SearchView;
import android.widget.Toast;

import java.util.Date;

import butterknife.BindColor;
import butterknife.BindView;
import butterknife.ButterKnife;
import io.github.hyuwah.catatanku.adapter.NoteCursorAdapter;
import io.github.hyuwah.catatanku.chrome.CustomTabActivityHelper;
import io.github.hyuwah.catatanku.storage.NoteContract;

public class NoteListActivity extends AppCompatActivity implements
    LoaderManager.LoaderCallbacks<Cursor> {

  // Views
  @BindView(R.id.fab)
  FloatingActionButton fab;
  @BindView(R.id.lv_note_list)
  ListView lvNoteList;
  @BindView(R.id.empty_note_list_view)
  View lvEmptyNoteList;


  // Debugging variable
  private Toast mToast;
  private int dummyDataCount = 10;

  // Storage
  private static final int NOTE_LOADER = 0;
  NoteCursorAdapter noteCursorAdapter;

  /**
   * Lifecycle Override
   */

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_note_list);

    ButterKnife.bind(this);

    fab.setOnClickListener(view -> {
      Intent intent = new Intent(NoteListActivity.this, EditorActivity.class);
      startActivity(intent);
    });

    prepareListView();

    getLoaderManager().initLoader(NOTE_LOADER, null, this);

  }

  @Override
  protected void onStart() {
    super.onStart();

  }

  /**
   * Overflow Menu Related
   */

  @Override
  public boolean onPrepareOptionsMenu(Menu menu) {
    super.onPrepareOptionsMenu(menu);
    MenuItem insertDummyData = menu.findItem(R.id.action_insert);
    insertDummyData.setTitle("Insert " + String.valueOf(dummyDataCount) + " data");
    return true;
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    // Inflate the menu; this adds items to the action bar if it is present.
    getMenuInflater().inflate(R.menu.menu_note_list, menu);

    //Associate searchable config with the Searchview
    SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
    SearchView searchView = (SearchView) menu.findItem(R.id.action_search).getActionView();
    searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));

    searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
      @Override
      public boolean onQueryTextSubmit(String s) {
        return false;
      }

      @Override
      public boolean onQueryTextChange(String s) {

        String selectionClause = NoteContract.NotesEntry.COLUMN_NOTE_TITLE + " LIKE ? OR "
            + NoteContract.NotesEntry.COLUMN_NOTE_BODY + " LIKE ?";
        String[] selectionArgs = new String[]{
            "%" + s + "%",
            "%" + s + "%"
        };

        Cursor cursor = getContentResolver().query(
            NoteContract.NotesEntry.CONTENT_URI,
            NoteContract.NotesEntry.DEFAULT_PROJECTION,
            selectionClause,
            selectionArgs,
            NoteContract.NotesEntry.SORT_TIME_DESC
        );
        Log.i(this.getClass().getSimpleName(), "onQueryTextChange: " + cursor.getCount());
        if (cursor != null) {
          noteCursorAdapter.swapCursor(cursor);
        }

        return false;
      }
    });

    return true;
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.action_delete_all:
        int rowsDeleted = getContentResolver()
            .delete(NoteContract.NotesEntry.CONTENT_URI, null, null);
        return true;

      case R.id.action_insert:
        generateDummyNote();
        return true;

      case R.id.action_gitbook_journal:
        openGitbookJournal();
        return true;

      case R.id.action_about:
        Intent intent = new Intent(this, AboutActivity.class);
        startActivity(intent);
        return true;

      case R.id.action_search:
        SearchView searchView = (SearchView) item.getActionView();
        searchView.setFocusable(true);
        searchView.setQueryHint("Text to search");
        searchView.setIconified(false);
        searchView.requestFocusFromTouch();
        return false;

      default:
        return super.onOptionsItemSelected(item);
    }
  }

  /**
   * Implements CursorLoader
   */

  @Override
  public Loader<Cursor> onCreateLoader(int i, Bundle bundle) {

    return new CursorLoader(this, NoteContract.NotesEntry.CONTENT_URI,
        NoteContract.NotesEntry.DEFAULT_PROJECTION,
        null,
        null,
        NoteContract.NotesEntry.SORT_TIME_DESC);
  }

  @Override
  public void onLoadFinished(Loader<Cursor> loader, Cursor cursor) {
    noteCursorAdapter.swapCursor(cursor);
  }

  @Override
  public void onLoaderReset(Loader<Cursor> loader) {
    noteCursorAdapter.swapCursor(null);
  }

  /**
   * Activity Methods
   */

  // ListView Related
  private void prepareListView() {

    noteCursorAdapter = new NoteCursorAdapter(this, null);

    lvNoteList.setAdapter(noteCursorAdapter);
    lvNoteList.setEmptyView(lvEmptyNoteList);

    // Multi select listview
    lvNoteList.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE_MODAL);
    lvNoteList.setMultiChoiceModeListener(new AbsListView.MultiChoiceModeListener() {
      @Override
      public void onItemCheckedStateChanged(ActionMode actionMode, int i, long l, boolean b) {
        final int checkedCount = lvNoteList.getCheckedItemCount();
        actionMode.setTitle(checkedCount + " Selected");
        noteCursorAdapter.toggleSelection(i);
      }

      @Override
      public boolean onCreateActionMode(ActionMode actionMode, Menu menu) {
        actionMode.getMenuInflater().inflate(R.menu.menu_note_list_onselect, menu);
        return true;
      }

      @Override
      public boolean onPrepareActionMode(ActionMode actionMode, Menu menu) {
        return false;
      }

      // ActionMode Menu
      @Override
      public boolean onActionItemClicked(ActionMode actionMode, MenuItem menuItem) {

        switch (menuItem.getItemId()) {

          case R.id.onselect_select_all:

            noteCursorAdapter.removeSelection();
            for (int x = 0; x < noteCursorAdapter.getCount(); x++) {
              lvNoteList.setItemChecked(x, true);
            }
            final int checkedCount = lvNoteList.getCheckedItemCount();
            actionMode.setTitle(checkedCount + " Selected");

            return true;
          case R.id.onselect_delete:

            showDeleteConfirmationDialog((dialogInterface, i) -> {
              for (int x = 0; x < noteCursorAdapter.getSelectedCount(); x++) {
                int listId = noteCursorAdapter.getSelectedIds().keyAt(x);
                long realId = noteCursorAdapter.getItemId(listId);
                Log.i(this.getClass().getSimpleName(),
                    "Selected: list id=" + listId + ", db_id=" + realId);
                int rowsDeleted = getContentResolver()
                    .delete(ContentUris.withAppendedId(NoteContract.NotesEntry.CONTENT_URI, realId),
                        null, null);
                Log.i(this.getClass().getSimpleName(),
                    "onActionItemClicked: rowsDeleted=" + rowsDeleted);

              }
              Toast.makeText(NoteListActivity.this,
                  "Deleted " + noteCursorAdapter.getSelectedCount() + " notes", Toast.LENGTH_SHORT)
                  .show();
              noteCursorAdapter.removeSelection();
              actionMode.finish();
            });

            return true;
          default:

            return false;

        }
      }

      @Override
      public void onDestroyActionMode(ActionMode actionMode) {
        noteCursorAdapter.removeSelection();
      }
    });

    // Onclick listView
    lvNoteList.setOnItemClickListener((adapterView, view, position, id) -> {
      Intent intent = new Intent(NoteListActivity.this, EditorActivity.class);
      Uri currentNoteUri = ContentUris.withAppendedId(NoteContract.NotesEntry.CONTENT_URI, id);
      intent.setData(currentNoteUri);
      startActivity(intent);
    });

  }

  private void showDeleteConfirmationDialog(DialogInterface.OnClickListener deleteClickListener) {

    AlertDialog.Builder builder = new AlertDialog.Builder(this);
    builder.setMessage("Selected note(s) will be deleted!")
        .setPositiveButton("Cancel", (dialogInterface, i) -> {
          if (dialogInterface != null) {
            dialogInterface.dismiss();
          }
        })
        .setNegativeButton("Delete", deleteClickListener).show();
  }

  private void openGitbookJournal() {
    String url = "https://hyuwah.gitbooks.io/journal-refactory/content/";
    Uri uri = Uri.parse(url);

    CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
    builder.setToolbarColor(ContextCompat.getColor(this, R.color.colorPrimary));
    builder.addDefaultShareMenuItem();
    builder.setShowTitle(true);

    CustomTabsIntent customTabsIntent = builder.build();
    CustomTabActivityHelper.openCustomTab(this, customTabsIntent, uri,
        (activity, uri1) -> {
          Intent intent = new Intent(Intent.ACTION_VIEW, uri1);
          activity.startActivity(intent);
        });
  }

  /**
   * Dummy notes data
   */

  private void generateDummyNote() {

    for (int i = 0; i < dummyDataCount; i++) {

      int randomTitleNum = (int) Math.floor(Math.random() * 100);
      int randomBodyNum = (int) Math.floor(Math.random() * 1000);

      ContentValues values = new ContentValues();
      values.put(NoteContract.NotesEntry.COLUMN_NOTE_TITLE, "Judul" + randomTitleNum);
      values.put(NoteContract.NotesEntry.COLUMN_NOTE_BODY,
          randomBodyNum + ". Lorem ipsum dolor sit amet");
      values.put(NoteContract.NotesEntry.COLUMN_NOTE_DATETIME, new Date().getTime());

      Uri newUri = getContentResolver().insert(NoteContract.NotesEntry.CONTENT_URI, values);
      Log.i(this.getClass().getSimpleName(), "generateDummyNote: " + values.toString());
    }

  }

}

File: app/src/main/java/io/github/hyuwah/catatanku/ui/notelist/NoteListActivity.kt
package io.github.hyuwah.catatanku.ui.notelist

import android.app.Activity
import android.app.AlertDialog
import android.app.SearchManager
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import dagger.hilt.android.AndroidEntryPoint
import io.github.hyuwah.catatanku.R
import io.github.hyuwah.catatanku.databinding.ActivityNoteListBinding
import io.github.hyuwah.catatanku.domain.model.Note
import io.github.hyuwah.catatanku.ui.about.AboutActivity
import io.github.hyuwah.catatanku.ui.editor.EditorActivity
import io.github.hyuwah.catatanku.utils.adjustInsets
import io.github.hyuwah.catatanku.utils.chrome.CustomTabActivityHelper

@AndroidEntryPoint
class NoteListActivity : AppCompatActivity(), NotesAdapter.NoteItemListener {
    
    private lateinit var binding: ActivityNoteListBinding

    private val viewModel: NoteListViewModel by viewModels()

    private var mBackPressed: Long = 0

    private val notesAdapter by lazy {
        NotesAdapter(noteItemListener = this)
    }

    private val sharedPref by lazy {
        getSharedPreferences(getString(R.string.pref_file_key), MODE_PRIVATE)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityNoteListBinding.inflate(layoutInflater)
        setContentView(binding.root)
        adjustInsets()
        setSupportActionBar(binding.toolbar)
        binding.fab.setOnClickListener {
            startActivity(Intent(this, EditorActivity::class.java))
        }
        prepareNoteList()
    }

    override fun onStart() {
        super.onStart()
        invalidateOptionsMenu()
    }

    override fun onBackPressed() {
        if (mBackPressed + TIME_INTERVAL > System.currentTimeMillis()) {
            super.onBackPressed()
            return
        } else {
            Toast.makeText(this, "Tap back again to exit", Toast.LENGTH_SHORT).show()
        }
        mBackPressed = System.currentTimeMillis()
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        super.onPrepareOptionsMenu(menu)
        return true
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_note_list, menu)

        //Associate searchable config with the Searchview
        val searchManager = getSystemService(SEARCH_SERVICE) as SearchManager
        val searchView = menu.findItem(R.id.action_search).actionView as? SearchView
        searchView?.setSearchableInfo(searchManager.getSearchableInfo(componentName))
        searchView?.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean {
                return false
            }

            override fun onQueryTextChange(s: String): Boolean {
                viewModel.search(s)
                return false
            }
        })

        // Check SharedPreferences for Overflow Menu
        val debugToggle = sharedPref.getBoolean(getString(R.string.pref_key_isdebug), false)
        menu.findItem(R.id.action_group_debug).isVisible = debugToggle
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_delete_all -> {
                viewModel.deleteAll()
                true
            }

            R.id.action_insert -> {
                viewModel.debugInsert()
                true
            }

            R.id.action_gitbook_journal -> {
                openGitbookJournal()
                true
            }

            R.id.action_about -> {
                val intent = Intent(this, AboutActivity::class.java)
                startActivity(intent)
                true
            }

            R.id.action_search -> {
                (item.actionView as SearchView?)?.run {
                    isFocusable = true
                    queryHint = "Text to search"
                    isIconified = false
                    requestFocusFromTouch()
                }
                false
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun adjustInsets() {
        with(binding) {
            fab.adjustInsets(navigationBar = true)
            rvNoteList.adjustInsets(navigationBar = true)
        }
    }

    private fun prepareNoteList() {
        viewModel.notesLiveData.observe(this) { notes ->
            binding.emptyNoteListView.isVisible = notes.isEmpty()
            notesAdapter.submitList(notes.toList())
        }
        binding.rvNoteList.adapter = notesAdapter
    }

    override fun onNoteClicked(note: Note) {
        val intent = Intent(this, EditorActivity::class.java).apply {
            putExtra("ID", note.id)
        }
        startActivity(intent)
    }

    override fun onNoteLongClicked(note: Note) {
        showDeleteConfirmationDialog(
            onDeleteClicked = { viewModel.deleteById(note.id) }
        )
    }

    private fun showDeleteConfirmationDialog(onDeleteClicked:() -> Unit) {
        AlertDialog.Builder(this)
            .setMessage("Selected note(s) will be deleted!")
            .setPositiveButton("Cancel") { dialogInterface: DialogInterface?, _: Int ->
                dialogInterface?.dismiss()
            }
            .setNegativeButton("Delete") { _: DialogInterface?, _: Int ->
                onDeleteClicked.invoke()
            }
            .show()
    }

    private fun openGitbookJournal() {
        val url = "https://hyuwah.gitbooks.io/journal-refactory/content/"
        val uri = Uri.parse(url)
        val customTabIntent = CustomTabsIntent.Builder()
            .setToolbarColor(ContextCompat.getColor(this, R.color.colorPrimary))
            .setShareState(CustomTabsIntent.SHARE_STATE_ON)
            .setShowTitle(true)
            .build()
        CustomTabActivityHelper.openCustomTab(this, customTabIntent, uri) { activity: Activity, uri1: Uri? ->
            activity.startActivity(Intent(Intent.ACTION_VIEW, uri1))
        }
    }

    companion object {
        // Double tap back to exit
        private const val TIME_INTERVAL =
            2000 // # milliseconds, desired time passed between two back presses.
    }
}